File: structural/dart/proxy_structured/remote_proxy.dart
abstract class RemoteServer {
  void request(String data);
}

class RealRemoteServer extends RemoteServer {
  @override
  void request(String data) {
    print('Remote server processing request: $data');
  }
}

class RemoteServerProxy extends RemoteServer {
  RealRemoteServer? _remoteServer;
  final String serverAddress;
  
  RemoteServerProxy(this.serverAddress);
  
  void _connectToServer() {
    print('Connecting to remote server at $serverAddress...');
  }
  
  @override
  void request(String data) {
    if (_remoteServer == null) {
      _connectToServer();
      _remoteServer = RealRemoteServer();
    }
    print('Proxy forwarding request to remote server...');
    _remoteServer!.request(data);
  }
}

File: structural/typescript/proxy_structured/remote_proxy.ts
export interface RemoteServer {
  request(data: string): void;
}

export class RealRemoteServer implements RemoteServer {
  request(data: string): void {
    console.log(`Remote server processing request: ${data}`);
  }
}

export class RemoteServerProxy implements RemoteServer {
  private remoteServer?: RealRemoteServer;
  private serverAddress: string;
  
  constructor(address: string) {
    this.serverAddress = address;
  }
  
  private connectToServer(): void {
    console.log(`Connecting to remote server at ${this.serverAddress}...`);
  }
  
  request(data: string): void {
    if (!this.remoteServer) {
      this.connectToServer();
      this.remoteServer = new RealRemoteServer();
    }
    console.log('Proxy forwarding request to remote server...');
    this.remoteServer.request(data);
  }
}
